server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionPluginDownloader.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.plugins.edition;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Set;
+import org.sonar.api.utils.HttpDownloader;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.core.util.FileUtils;
+import org.sonar.server.platform.ServerFileSystem;
+import org.sonar.updatecenter.common.Release;
+import org.sonar.updatecenter.common.UpdateCenter;
+import org.sonar.updatecenter.common.Version;
+
+import static org.apache.commons.io.FileUtils.toFile;
+import static org.apache.commons.lang.StringUtils.substringAfterLast;
+
+public class EditionPluginDownloader {
+  private static final Logger LOG = Loggers.get(EditionPluginDownloader.class);
+  private static final String PLUGIN_EXTENSION = "jar";
+
+  private final Path tmpDir;
+  private final Path downloadDir;
+  private final HttpDownloader downloader;
+
+  public EditionPluginDownloader(HttpDownloader downloader, ServerFileSystem fileSystem) {
+    this.downloadDir = fileSystem.getEditionDownloadedPluginsDir().toPath();
+    this.downloader = downloader;
+    this.tmpDir = downloadDir.resolveSibling(downloadDir.getFileName() + "_tmp");
+  }
+
+  public void downloadEditionPlugins(Set<String> pluginKeys, UpdateCenter updateCenter) {
+    try {
+      Set<Release> pluginsToInstall = new HashSet<>();
+      for (String pluginKey : pluginKeys) {
+        pluginsToInstall.addAll(updateCenter.findInstallablePlugins(pluginKey, Version.create("")));
+      }
+
+      FileUtils.deleteDirectory(tmpDir);
+      Files.createDirectories(tmpDir);
+
+      for (Release r : pluginsToInstall) {
+        download(r);
+      }
+
+      FileUtils.deleteDirectory(downloadDir);
+      Files.move(tmpDir, downloadDir);
+    } catch (IOException e) {
+      throw new IllegalStateException(e.getMessage(), e);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00112)